### PR TITLE
Add failure handling

### PR DESCRIPTION
**Pull request by klippx**
This is a preview of upcoming changes to do error handling. Please add comments.

FYI: The Events API changes are rudimentary and not really started.

### CODE REVIEW

File: lib/phobos_db_checkpoint/failure.rb [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    def payload
+      attributes['payload'].deep_symbolize_keys
+    end
+
+    def metadata
+      attributes['metadata'].deep_symbolize_keys
+    end
+
+    # Can we delete the failure already at this stage?
```
Comment (klippx) [inline lib/phobos_db_checkpoint/failure.rb]: Thoughts about this are welcome. Like I write, I am thinking retry! should delete the record. It will be  created again if it winds up failing enough times again.

File: lib/phobos_db_checkpoint/handler.rb [hunk 2]
```diff
@@ -14,6 +15,19 @@ module ClassMethods
...
+      # Name...
+      # => retry_consume?
+      # => back_off?
+      # => consume_again?
+      # => fail_consume?
+      # => fail_permanently?
+      # => give_up?
+      # => stop_message?
+      # => consume_failure_condition_met?
+      def retry_consume?(event, event_metadata, exception)
```
Comment (klippx) [inline lib/phobos_db_checkpoint/handler.rb]: Actually I think this name is kind of decent now. Read it from a readme point of view, or a KlearSearch point of view, it seems like a name that makes sense

File: lib/phobos_db_checkpoint/handler.rb [hunk 3]
```diff
@@ -31,7 +45,15 @@ def around_consume(payload, metadata)
           end
 
           event_action = instrument('db_checkpoint.event_action', event_metadata) do
-            yield
+            begin
```
Comment (klippx) [inline lib/phobos_db_checkpoint/handler.rb]: I know @ticolucci has comments about this being a separate handler.

But I think error handling is mandatory, less classes and less confusion. You could always configure retry_consume? to always return true for example, to escape this functionality and restore the behaviour before having error handling.

WDYT?

File: lib/phobos_db_checkpoint/events_api.rb [hunk 4]
```diff
@@ -72,5 +72,10 @@ class EventsAPI < Sinatra::Base
         .offset(offset)
         .to_json
     end
+
+    get "/#{VERSION}/error_events" do
+      content_type :json
+      PhobosDBCheckpoint::Failure.all.to_json
```
Comment (tulios) [inline lib/phobos_db_checkpoint/events_api.rb]: I think this should be paginated, it's easy to end up with thousands of failures when something happen

File: lib/phobos_db_checkpoint/failure.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    def retry!
+      configured_handler
+        .new
+        .consume(
+          payload,
+          metadata.merge(retry_count: 0)
+        )
+    end
+
+    def configured_handler
```
Comment (tulios) [inline lib/phobos_db_checkpoint/failure.rb]: Should we try to unify this implementation in a module? It's pretty much the same as https://github.com/klarna/phobos_db_checkpoint/blob/02381bf8fea5935048977e24931242aeb1e40ea1/lib/phobos_db_checkpoint/event.rb#L17

This one is actually better
WDYT?

File: lib/phobos_db_checkpoint/handler.rb (hunk 3 above)
Comment (tulios) [inline lib/phobos_db_checkpoint/handler.rb]: We can add a new section to phobos.yml with the `max_retries`, something like:

```yaml
db_checkpoint:
  max_retries: 3
```

If the attribute is `nil` we keep the old behavior as the default, we keep retrying forever. If it's configured we apply the new rule, saving the failure after X retries.

WDYT? @ticolucci

File: spec/lib/phobos_db_checkpoint/failure_spec.rb [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    it 'stores payload as json' do
+      expect(subject.attributes['payload']).to eql JSON(event_payload.to_json)
+    end
+
+    it 'stores metadata as json' do
+      expect(subject.attributes['metadata']).to eql JSON(event_metadata.to_json)
+    end
+
+    context 'with exception' do
+      let(:exception) do
```
Comment (tulios) [inline spec/lib/phobos_db_checkpoint/failure_spec.rb]: This should work:

```ruby
0/0 rescue e
```

File: spec/lib/phobos_db_checkpoint/handler_spec.rb [hunk 7]
```diff
@@ -3,22 +3,22 @@
 require 'json'
 
 RSpec.describe PhobosDBCheckpoint::Handler, type: :db do
-  class TestPhobosDbCheckpointHander
+  class TestPhobosDbCheckpointHandler
```
Comment (tulios) [inline spec/lib/phobos_db_checkpoint/handler_spec.rb]: Ops... my bad :p

File: lib/phobos_db_checkpoint/handler.rb (hunk 3 above)
Comment (klippx) [inline lib/phobos_db_checkpoint/handler.rb]: Your suggestion basically boils down to some derivative of this sample code below:

```ruby
      def retry_consume?(event, event_metadata, exception)
        return false unless Phobos.config&.db_checkpoint&.max_retries
        event_metadata[:retry_count] < Phobos.config&.db_checkpoint&.max_retries
      end
```

Your suggestion is that if max_retries isn't set at all or set to nil, then the expected result is keep retrying forever...

However, is that very useful outside of implementations using Phobos that handles errors themselves? 
Would a vanilla Phobos+PhobosDBCheckpoint user expect an event to "get stuck" and fail in a never ending loop?

From this point of view I must say I am more strongly in favor of that if you haven't configured a max retry it should fail immediately instead.

Comment (ticolucci) [inline lib/phobos_db_checkpoint/handler.rb]: I like @tulios solution. It keeps the old behaviour and the example @klippx brought is very clean.

TL;DR - it boils down to which of the following costs more:
* consuming out-of order, but increasing throughput
* ensuring the order, but stoping the process

For me it does make sense to block a kafka partition on errors by default. Remember that blocking a Phobos handler forever actually means blocking one partition, not the entire process. Since the main point for many Kafka usages is that order is guaranteed by partition, if my consumer marks a message as failed, continues the process and then enters an unexpected state because we consumed messages out of order this might cause really huge losses, where a blocked consumer will likely cause frustration.

Comment (ticolucci) [inline lib/phobos_db_checkpoint/handler.rb]: Mikael brought an interesting point. Maybe we don't have to default. =P

If the user needs to make a conscious choice about this property, then there will be no surprises. So, in the future, the consumer groups could break and not start at all if the config is not set.

For now, I agree with @klippx idea of: when the config is not set, default to the old behaviour (infinity retries) _but_ warns the user about the "deprecation" (or future requirement) of explicitly choosing:

```yaml
listeners:
  - handler: MyHandler
    topic: le_topic
    # and then either
    max_retries: 3 # the handler will retry this many times each message before storing them in the quarantine Failure table
    # or
    retry_forever: true
```

Comment (tulios) [inline lib/phobos_db_checkpoint/handler.rb]: Exactly, we can't take for granted that a message can be skipped and the next will succeed. It can have requirements on the previous message, it's up to the developer to configure phobos to fits its needs.

Comment (tulios) [inline lib/phobos_db_checkpoint/handler.rb]: @ticolucci I think this adds complexity. PhobosDBCheckpoint is a plugin for Phobos, if not configured it should keep the original behavior, which is retry forever.

File: templates/migrate/phobos_02_create_failures.rb.erb [hunk 8]
```diff
@@ -0,0 +1,16 @@
+class Phobos02CreateFailures < ActiveRecord::Migration[<%= ActiveRecord::Migration.current_version %>]
+  def up
+    create_table :failures do |t|
+      t.json      :payload,        index: false
+      t.json      :metadata,       index: false
+      t.string    :checksum,       index: true
+      t.string    :error_class,    index: false
```
Comment (tulios) [inline templates/migrate/phobos_02_create_failures.rb.erb]: Maybe index true on `error_class`, it might be useful.

File: lib/phobos_db_checkpoint/event_helper.rb [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+  module EventHelper
+    def configured_handler
+      Phobos
+        .config
+        .listeners
+        .find { |l| l.group_id == self.group_id }
+        .handler
+        .constantize
+    rescue NoMethodError => e
+      raise(HandlerNotFoundError, self.group_id) if e.message =~ /handler/
```
Comment (ticolucci) [inline lib/phobos_db_checkpoint/event_helper.rb]: if we split the calls into `listener = ...` then try `listener.handler.constantize` we can do the nil check and raise in between.
I prefer that over trying to see which method was not found.

File: lib/phobos_db_checkpoint/handler.rb (hunk 3 above)
Comment (ticolucci) [inline lib/phobos_db_checkpoint/handler.rb]: ok. good point about dbCheckpoint being a plugin.

@klippx , do you also agree? do we have a verdict =P?

File: lib/phobos_db_checkpoint/failure.rb [hunk 10]
```diff
@@ -0,0 +1,49 @@
+module PhobosDBCheckpoint
+  class Failure < ActiveRecord::Base
+    include PhobosDBCheckpoint::EventHelper
+
+    scope :by_checksum, -> (val) { where("metadata->>'checksum' = ?", val) }
+    scope :by_topic, -> (val) { where("metadata->>'topic' = ?", val) }
```
Comment (ticolucci) [inline lib/phobos_db_checkpoint/failure.rb]: We chat offline already, so just bringing the discussion online.

We think we cannot use the json queries, since it's not supported by all DBs. Therefore we concluded it's better to create columns and indexes for `topic` and `group_id` so that we can filter failures later on.

File: README.md [hunk 11]
```diff
@@ -119,6 +120,33 @@ If your handler returns anything different than an __ack__ it won't be saved to
...
+If your handler fails during the process of consuming the event, the event will be processed again later since it is neither acknowledged or skipped. This could go on indefinitely, so in order to help you deal with this PhobosDBCheckpoint will after a configurable number of consecutive attempts mark them as permanently failed. This config is set in Phobos configuration as such:
+
+```yml
+db_checkpoint:
+  max_retries: 3
+```
+
+The retry decision is driven by inspecting the retry counter in the Phobos metadata, and if not meeting the retry criteria it will result in creating a `Failure` record and then skipping the event. You can easily retry these events later by simply invoking `retry!` on them.
+
+*Nota bene:* If not configured, `PhobosDBCheckpoint::Handler` will fail events upon first encountered exception.
```
Comment (tulios) [inline README.md]: Remember to update this after our discussions

File: lib/phobos_db_checkpoint/handler.rb (hunk 3 above)
Comment (klippx) [inline lib/phobos_db_checkpoint/handler.rb]: Lets go with the keep old behaviour. I skipped adding warnings for now, so retry forever sounds good

File: templates/migrate/phobos_02_create_failures.rb.erb [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+      t.string    :event_type,     index: true
+      t.string    :event_version,  index: true
+      t.string    :checksum,       index: true
+      t.json      :payload,        index: false
+      t.json      :metadata,       index: false
+      t.string    :error_class,    index: false
+      t.string    :error_message,  index: false
+      t.json      :error_backtrace
+    end
+  end
```
Comment (ticolucci) [inline templates/migrate/phobos_02_create_failures.rb.erb]: I think we need at least indices for `created_at`, `event_type` and `event_version`. Since they are the ones I see myself querying by to debug. 
Thoughts?

Comment (ticolucci) [inline templates/migrate/phobos_02_create_failures.rb.erb]: Douh! https://4.bp.blogspot.com/_CdjH-c4g0mI/ST6EUU_aFII/AAAAAAAAAnY/QFROKWIc5-M/s320/doh_l-761608.gif

Missed the bunch `index: true` on the migration... forget about htis.. carry on =)
